test/state/mpt.cpp

```diff
@@ -0,0 +1,283 @@
+    {
+    case Kind::branch:
+    {
+        assert(m_path.length == 0);  // Branch has no path.
+
+        const auto idx = path.nibbles[0];
+        auto& child = m_children[idx];
+        if (!child)
+            child = std::make_unique<MPTNode>(leaf(path.tail(1), std::move(value)));
+        else
+            child->insert(path.tail(1), std::move(value));
+        break;
+    }
+
+    case Kind::ext:
+    {
+        assert(m_path.length != 0);  // Ext must have non-empty path.
+
+        const auto mismatch_pos = mismatch(m_path, path);
+
+        if (mismatch_pos == m_path.length)
+        {
+            // Paths match: go into the child.
+            return m_children[0]->insert(path.tail(mismatch_pos), std::move(value));
+        }
+
+        const auto orig_idx = m_path.nibbles[mismatch_pos];
+        const auto new_idx = path.nibbles[mismatch_pos];
+
+        auto br = branch(orig_idx,
+            (mismatch_pos < m_path.length - 1) ?
+                std::make_unique<MPTNode>(
+                    ext(m_path.tail(mismatch_pos + 1), std::move(m_children[0]))) :
+                std::move(m_children[0]),
+            new_idx,
+            std::make_unique<MPTNode>(leaf(path.tail(mismatch_pos + 1), std::move(value))));
+
+        *this = (mismatch_pos == 0) ?
+                    std::move(br) :
+                    ext(m_path.head(mismatch_pos), std::make_unique<MPTNode>(std::move(br)));
+        break;
+    }
+
+    case Kind::leaf:
+    {
+        assert(m_path.length != 0);  // Leaf must have non-empty path.
+
+        const auto mismatch_pos = mismatch(m_path, path);
+        assert(mismatch_pos != m_path.length);  // Paths must be different.
+
+        const auto orig_idx = m_path.nibbles[mismatch_pos];
+        const auto new_idx = path.nibbles[mismatch_pos];
+
+        auto br = branch(orig_idx,
+            std::make_unique<MPTNode>(leaf(m_path.tail(mismatch_pos + 1), std::move(m_value))),
+            new_idx,
+            std::make_unique<MPTNode>(leaf(path.tail(mismatch_pos + 1), std::move(value))));
+
+        *this = (mismatch_pos == 0) ?
+                    std::move(br) :
+                    ext(m_path.head(mismatch_pos), std::make_unique<MPTNode>(std::move(br)));
+        break;
+    }
+
+    default:
+        assert(false);
+    }
+}
+
+hash256 MPTNode::hash() const  // NOLINT(misc-no-recursion)
+{
+    hash256 r{};
+    switch (m_kind)
+    {
+    case Kind::leaf:
+    {
+        const auto node = rlp::encode_tuple(m_path.encode(false), m_value);
+        r = keccak256(node);
+        break;
+    }
+    case Kind::branch:
+    {
+        assert(m_path.length == 0);
+
+        // Temporary storage for children hashes.
+        // The `bytes` type could be used instead, but this way dynamic allocation is avoided.
+        hash256 children_hashes[num_children];
+
+        // Views of children hash bytes. Additional item for hash list
+        // terminator (always empty). Does not seem needed for correctness,
```

Comment:
weird how it could not influence correctness, hash would be different without it?